designs/CB-OPT-design-week7-update.md

```diff
@@ -0,0 +1,224 @@
+
+We have implemented a majority of our design decisions in "optimize_ast.ts". We have implemented dead code elimination, including unreachable instructions after `return` and dead branches such as `if True` and `while False`. Our program can also do constant folding for all operations whose operands are constant. Further folding will be available when the constant propagation is facilitated.
+
+The optimization takes place in a iterative fashion until the source code remains unchanged during a pass. Currently, passes available are:
+
+* AST-Level: 
+  1. dead code elimination
+  2. constant folding
+* IR-Level:
+  1. constant folding
+
+
+## D. New decisions
+We plan to implement the following optimizations in the next sprint: 1. Constant propagation 2. Liveness Analysis. 
+
+Most likely, we will try to use control-flow anlyasis metioned in class for our implementation on constant propagation which indicates that we will implement constant propagtaion in "optimize_ir.ts" first. 
+
+We also plan to implement liveness analysis at `ir` first since the control flow at this level is less complex, but it does not preclude our intention to introduce this analysis at `ast` since removing dead variable at the earlier stage may bring more benefits, and we can exploit more control-flow semantics despite its complexity.
+
+We already observe noticable difference in the optimality of enforcing the same optimization techniques at different levels. For constant folding, if we can merge operations at `ast`, less intermediate variable will be generated compared with applying folding at `ir` level. Although these redundant variables may be eliminated by further propagation and folding at `ir`, wipping the **seed** out at `ast` can definitely simplify the further optimization. 
+
+## E. Test
+To conduct automatic testing, we implement two functions in the `./tests/asserts.test.ts`, which are `assertOptimize` and `assertPass` functions. For the `assertOptimize` function, we run the codes before and after optimization respectively. **When the runtime results of two generated WASM codes matched, and the length of the optimized WASM code is shorter than the original WASM codes, we consider a test case is passed.** Similarly, for `assertPass` function, the only difference with `assertOptimize` is that the test case is considered pass when the length of optimized WASM code is shorter or equal to the original WASM code. The purpose for this function is for the test cases of `pass` statements since such statements do not generate any WASM codes. For example, a test case like,
```

Comment:
I have a suggestion, if not too much trouble, try to have flags in the compiler to turn different types of optimization on and off. Then run your tests gradually. I suspect you'll run into failures and in that case, you can at least narrow down which layer of optimization is causing the issue and debug faster.